### DISCUSSION TITLE
Startup script conflicts with qnap conventions

### DISCUSSION THREAD

**Discussion by ittayd** [Entware/Entware#774] [Q&A] [answered]
The startup script deletes /opt. But some other packages (e.g. from Qoolbox in qnapclub) link from /opt/<pkg> to the .qpkg directory. And others rely on this (for example, the qbitorrent package adds /opt/QPython3 to the qbittorrent path so it can find python3).

**Comment by zyxmon** [comment, marked as answer]
Entware uses the same approach as Optware. Optware qpkg was developed long before Qoolbox packages. I can suggest other packages not to use /opt. /opt folder is essential for Optware and Entware and can be replaced by any other folder in other packages that use /opt.

Entware has python3 packages and tools to build own python packages. QPython3 is not needed wahe Entware is installed.

**Comment by ittayd** [comment]
QPython3 is required so that qBitorrent finds python3. Right now, the fact
that it deletes /opt means it is not usable. But moreover, it means it
impacts other packages.

**Comment by zyxmon** [comment]
Entware version of qbittorent uses python from entware.

**Comment by ittayd** [comment]
Well, yes, but it doesn't integrate into the appcenter, dashboard, etc.

Another thing is that there is a file /opt/nasconfig_fs.img.tgz which is
used by some init scripts. Not sure when or for what, but when /opt is
deleted, it is gone.

**Reply by ittayd** [reply]
used by init_disk.sh, killnas.sh (where, if it is not found the message "Error: DOM image crashed. (-3)" is printed), update_img.sh.

seems that a safer approach is to link just the subdirectories (/opt/bin etc.) and do it only if they don't already exist, and if they do, error and abort the installation.

**Reply by ittayd** [reply]
or just use something like /ent instead of /opt ...

**Reply by zyxmon** [reply]
>> file /opt/nasconfig_fs.img.tgz

/opt is located in tmpfs (in ram) and this file is recreated after reboot.

>> or just use something like /ent instead of /opt ...

Or just other packages should not use /opt which is essential for Optware and Entware.

**Reply by ittayd** [reply]
/opt existed from the days I was working on SGI Unix. Is there no way of making the prefix configurable so it can coexist with packages that are not aware that Entware exists?

**Reply by zyxmon** [reply]
1) There are no rules for directories that packages can use - this leads to conflicts! 
2) I have tried in 2004 to use /Apps/opt instead of /opt - https://forum.qnap.com/viewtopic.php?f=128&t=99489&hilit=zyxmon&start=30#p446371 - all the bugs found were because of this change.
3) In the time Entware was made for QNAP- no other packages used /opt but Optware that is now dead.
4) Entware is used on different routers, asusstor and synology NASes, android boxes,.... without this problem. I see not need to change /opt folder just for Qnap.
